### DISCUSSION TITLE
Using NLP parametric sensitivity for gradient descent

### DISCUSSION THREAD

**Discussion by aarya063** [casadi/casadi#3215] [Q&A]
Hello!

Considering the simple example of the Parametric sensitivity of NLP from the web.

opti = Opti();
x = opti.variable();
y = opti.variable();

xy = [x;y];
p = opti.parameter();

opti.minimize((1-x)^2 + 0.2*(y-x^2)^2);
opti.subject_to(x>=0);
opti.subject_to((p/2)^2 <= (x+0.5)^2+y^2 <= p^2);

opts = struct;
opts.ipopt.print_level = 0;
opts.print_time = false;
opti.solver('ipopt',opts);

M = opti.to_function('M',{p},{xy});

z = @(xy) xy(2,:)-xy(1,:);

pvec = linspace(1,2,100);
S = full(M(pvec));

opts = struct;
opts.qpsol = 'qrqp';
opti.solver('sqpmethod',opts);

Z = opti.to_function('Z',{p,xy},{z(xy)});

zp = Z(p,xy);
g = gradient(zp,p);

Suppose that I want to minimize z wrt p. I would like to know how I could use the gradient (zp,p) to give me a double matlab type and evaluate it for parameter values. 

It is clear that zp= Z(p,xy) is scalar-valued and so its gradient should be able to give a Matlab vector.

I guess my trouble is accessing and being able to use MX  in order to be able to perform the gradient descent step. 

Thank you for any kind of help 🙏 
Aarya

**Comment by yakunouyang** [comment]
I think you can just create a function and evaluate it:

```Matlab
G = Function('G', {p, xy}, {g});
p0 = 1.0;
result = G(p0, M(p0));
```
